ui/job-view/redux/stores/pinnedJobs.js

```diff
@@ -105,26 +106,37 @@ export const pinJobs = (jobsToPin) => {
   };
 };
 
-export const addBug = (bug, job) => {
+export const addBug = (bug, job = null) => {
   return async (dispatch, getState) => {
     const {
       pinnedJobs: { pinnedJobBugs, newBug },
     } = getState();
 
-    const bugId = bug.dupe_of ? bug.dupe_of : bug.id;
-    if (!pinnedJobBugs.has(bugId)) {
-      pinnedJobBugs.add(bugId);
-    }
-
+    const newBugUpdate = new Set(newBug);
     if ('newBug' in bug) {
       if (!newBug.has(bug.newBug)) {
-        newBug.add(bug.newBug);
+        newBugUpdate.add(bug.newBug);
       }
     }
+    // Avoid duplicating an already pinned bug
+    if (
+      pinnedJobBugs.some(
+        (b) =>
+          (b.dupe_of && (b.dupe_of === bug.id || b.dupe_of === bug.dupe_of)) ||
+          (b.id && (b.id === bug.id || b.id === bug.dupe_of)) ||
+          (b.internal_id && b.internal_id === bug.internal_id),
+      )
```

Comment:
this looks complicated and a high chance of error.

Reply:
The lines are separated with OR operators and each represents a subcheck:
* check if  bug already in pinboard is set as duplicate and that number matches either the bug number to be pinned or the bug it is as duplicate of
* check if a bug already in the pinboard has a number matching the number of the bug to be pinned or the number of the bug to which it is set as duplicate
* check if internal issue already in pinboard if classified with internal issue (assumption internal issue will be converted to bug soon enough to not need support for duplicates

Reply:
I'll add those as comments on each line so it is clear!
The idea is that the UI avoids sending duplicated values (even if the backend won't accept duplicates anyway). It was done in an easiest way before because using a Set containing IDs (or `dupe_of`).